cmd/zpool/zpool_main.c

```diff
@@ -2471,61 +2474,587 @@ zpool_do_import(int argc, char **argv)
+{
+	unsigned long column_width = 5; /* Normal niceprint */
+	static unsigned long widths[] = {
+		/*
+		 * Choose some sane default column sizes for printing the
+		 * raw numbers.
+		 */
+		[VS_DEFAULT] = 15, /* 1PB capacity */
+		[VS_LATENCY] = 10, /* 1B ns = 10sec */
+		[VS_QUEUES] = 6,   /* 1M queue entries */
+	};
+
+	if (cb->cb_literal)
+		column_width = widths[type];
+
+	return (column_width);
+}
+
+/*
+ * Utility function to print out a line of dashes like:
+ *
+ * 	--------------------------------  -----  -----  -----  -----  -----
+ *
+ * ...or a dashed named-row line like:
+ *
+ * 	logs                                  -      -      -      -      -
+ *
+ * @cb:				iostat data
+ *
+ * @force_column_width		If non-zero, use the value as the column width.
+ * 				Otherwise use the default column widths.
+ *
+ * @name:			Print a dashed named-row line starting
+ * 				with @name.  Otherwise, print a regular
+ * 				dashed line.
+ */
+static void
+print_iostat_dashes(iostat_cbdata_t *cb, unsigned int force_column_width,
+							const char *name)
+{
+	int i;
+	unsigned int dash_sets;
+	unsigned int namewidth;
+	uint64_t flags = cb->cb_flags;
+	uint64_t f;
+	int idx;
+
+	if (cb->cb_flags & VS_HISTO_M)
+		namewidth = MAX(cb->cb_namewidth, strlen("latency"));
+	else
+		namewidth = cb->cb_namewidth;
+
+	if (name) {
+		namewidth = MAX(cb->cb_namewidth, strlen(name));
+		printf("%-*s", namewidth, name);
+	} else {
+		for (i = 0; i < namewidth; i++)
+			(void) printf("-");
+	}
+
+	/* For each bit in flags */
+	for (f = flags; f; f &= ~BIT(idx)) {
+		unsigned int column_width;
+		idx = ffsll(f) - 1;
+		if (force_column_width)
+			column_width = force_column_width;
+		else
+			column_width = default_column_width(cb, idx);
+
+		/* Figure out how many sets of "-----" we need to print */
+		dash_sets = str_array_len(iostat_bottom_labels[idx]);
+
+		while (dash_sets--) {
+			if (name)
+				printf("  %*s-", column_width - 1, " ");
+			else
+				printf("  %.*s", column_width,
+						"--------------------");
+		}
+	}
+	printf("\n");
+}
+
+
+static void
+__print_iostat_separator(iostat_cbdata_t *cb, unsigned int force_column_width)
+{
+	print_iostat_dashes(cb, force_column_width, NULL);
+}
+
 static void
 print_iostat_separator(iostat_cbdata_t *cb)
 {
-	int i = 0;
+	__print_iostat_separator(cb, 0);
+}
+
+static void
+__print_iostat_header(iostat_cbdata_t *cb, unsigned int force_column_width,
+						const char *histo_vdev_name)
+{
+	int i, idx;
+	unsigned int text_start, rw_column_width;
+	unsigned int namewidth;
+	uint64_t flags = cb->cb_flags;
+	uint64_t f;
+	unsigned int column_width = force_column_width;
+
+	if (flags & VS_HISTO_M)
+		namewidth = MAX(cb->cb_namewidth, strlen("latency"));
+	else
+		namewidth = cb->cb_namewidth;
+
+	if (flags & VS_HISTO_M)
+		printf("%-*s", namewidth, histo_vdev_name);
+	else
+		printf("%*s", namewidth, "");
+
+	/* For each bit set in flags */
+	for (f = flags; f; f &= ~BIT(idx)) {
+		idx = ffsll(f) - 1;
```

Comment:
Instead of `ffsll()` which is a glibc extension use `highbit64()`.  This isn't as efficient but it is more portable, in general we try to avoid gnu specific extensions.  Related to this we could definitely update our `highbit64()` implementation in user space, which is ZoL specific, to use `ffsll()` internally.